bluetooth: shell: refactor shell print for Bluetooth-specific context

ndrs-pst opened the pull request:
This PR aims to eliminate the dependency on `ctx_shell` in the Bluetooth `host/shell/*`, making the code more maintainable.
- Introduced `bt_shell_private.c` and `bt_shell_private.h` to provide common functions for the Bluetooth shell.
- Limit the usage of `ctx_shell` to cases where printing requires it and `sh` is not available. 
(Interim commit before switch to `bt_shell_*`) 
- Refactor shell print to eliminate `ctx_shell` usage in the Bluetooth `host/shell/*`.

Fixes #41796

ndrs-pst committed:
bluetooth: shell: add `bt_shell_private.c` and `bt_shell_private.h`

Introduced `bt_shell_private.c` and `bt_shell_private.h` to provide
common functions for the Bluetooth `shell_wall_print`.
These functions are equivalent to `shell_fprintf`, `shell_info`,
`shell_print`, `shell_warn`, `shell_error` and `shell_hexdump`
but without requiring the `sh` parameter.

The cost of the newly added `bt_shell_fprintf_info` ... `_error` functions
will be negligible when there are many individual calls that need to pass
both the `sh` and `color` parameters each time.

Signed-off-by: Pisit Sawangvonganan <pisit@ndrsolution.com>

Thalley reviewed (dismissed):
I don't think this is the right direction (but please do convince me if you believe so). 

Some of the changes are welcome, but the idea of a `bt_shell` is IMO not. 

In your commit message you mention that you want to remove the code footprint, but you also mention an increase cost of this new bt_shell. Aren't those two sentences conflicting? 

The BT shell definitely needs improvement. I created a issue https://github.com/zephyrproject-rtos/zephyr/issues/70945 related to this, that requests the removed of the `ctx_shell`. The removal of the `ctx_shell` would directly conflict with the majority of this PR, as all the `bt_shell` functions you've defined will simply call the shell functions directly (as they would need the shell pointer again). 

Could you provide some data on the codesize that you've increase/decreased?

ndrs-pst commented:
@Thalley Thank you for your feedback. It's good to hear that you're addressing `ctx_shell`. 😃 

> Aren't those two sentences conflicting?

At first glance, this might seem controversial.
Currently, many shell print require passing parameters like `sh` and `color` repeatedly.
By centralizing these interactions within specific `bt_shell` functions, we streamline the code.

> Could you provide some data on the codesize that you've increase/decreased?

By picking up `periodic_adv_conn` sample with `nrf52840dk` boards and adding `CONFIG_BT_SHELL=y`
the footprint of `shell\*` which get from `rom_report` shown that 

| MODULE  | FLASH SIZE (BEFORE) | FLASH SIZE (AFTER) | INCREASE/DECREASE |
| ------------- | --------------------------- | ---------------------------------- | ---- |
| `bluetooth\shell\*`  | 15,795 B | 14,963 B | -832 B |
| `bt.c`  | 13,251 B | 12,275 B | -976 B |
| `bt_shell_private.c`  | 0 B | 144 B | +144 B |

Here is the footprint from each wrapper
![bt_shell_private](https://github.com/zephyrproject-rtos/zephyr/assets/50133174/b4e32f90-9b87-471d-9c3a-0e47c66b1428)

ndrs-pst commented:
To stretch further, I think the functions that should be improved are
`shell_fprintf_impl(sh, color, fmt, ##__VA_ARGS__)` itself to have a wrapper for the color parameter.

```
void __printf_like(2, 3) shell_fprintf_error_impl(const struct shell *sh, const char *fmt, ...);
void __printf_like(2, 3) shell_fprintf_info_impl(const struct shell *sh, const char *fmt, ...);
void __printf_like(2, 3) shell_fprintf_print_impl(const struct shell *sh, const char *fmt, ...);
void __printf_like(2, 3) shell_fprintf_warn_impl(const struct shell *sh, const char *fmt, ...);
```

Since in reality the same message printed out inside the shell is always fixed to a specific color, and if we prioritize the code size, this should make sense.

Otherwise, every time, for example in Cortex-M, `r1` needs to be allocated to pass the `color` value (while r0 is for `sh`), which causes the caller code to save the previously used `r1`.

Thalley commented:
Just tried your PR with the BT shell test application for the `nrf5340dk_nrf5340_cpuapp` board and got the following results:

With PR
```
Memory region         Used Size  Region Size  %age Used
           FLASH:      317356 B         1 MB     30.27%
             RAM:       52992 B       448 KB     11.55%
        IDT_LIST:          0 GB        32 KB      0.00%
```

Without PR
```
Memory region         Used Size  Region Size  %age Used
           FLASH:      318668 B         1 MB     30.39%
             RAM:       52992 B       448 KB     11.55%
        IDT_LIST:          0 GB        32 KB      0.00%
```

So while I do not agree with the direction of this PR, as we really need to get rid of `ctx_shell`, I can't disagree with the 1312 Bytes saved. 

I'm not at a point where I'll approve, but I'm also not going to reject this change. I'd like to get input from the other BT collaborators on this.

ndrs-pst commented:
@Thalley, thank you for the additional information. It seems we should start by getting rid of `ctx_shell` and then find a way to reduce the footprint.

ndrs-pst commented:
> To stretch further, I think the functions that should be improved are `shell_fprintf_impl(sh, color, fmt, ##__VA_ARGS__)` itself to have a wrapper for the color parameter.
> 
> ```
> void __printf_like(2, 3) shell_fprintf_error_impl(const struct shell *sh, const char *fmt, ...);
> void __printf_like(2, 3) shell_fprintf_info_impl(const struct shell *sh, const char *fmt, ...);
> void __printf_like(2, 3) shell_fprintf_print_impl(const struct shell *sh, const char *fmt, ...);
> void __printf_like(2, 3) shell_fprintf_warn_impl(const struct shell *sh, const char *fmt, ...);
> ```
> 
> Since in reality the same message printed out inside the shell is always fixed to a specific color, and if we prioritize the code size, this should make sense.
> 
> Otherwise, every time, for example in Cortex-M, `r1` needs to be allocated to pass the `color` value (while r0 is for `sh`), which causes the caller code to save the previously used `r1`.

@Thalley Just a reminder that this stretch was merged in this PR: https://github.com/zephyrproject-rtos/zephyr/pull/75340. 👀

ndrs-pst commented:
@Thalley I just added another commit that introduces `CONFIG_BT_SHELL_BACKEND` for Bluetooth shell output when the shell context is unavailable (e.g., in callbacks).

While exploring ways to eliminate `ctx_shell`, could we initialize it during startup and hide it from other shell commands?
Since `ctx_shell` is currently used in multiple places like `audio/shell`, `mesh/shell`, and `services/ias/shell`.

Thalley commented:
> While exploring ways to eliminate `ctx_shell`, could we initialize it during startup and hide it from other shell commands?

Question is, initialize it to what? 
The challenge (and main issue) with the `ctx_shell` is that it's not a constant. It's just a variable assigned to _some_ shell, of which there can be more 1 or more. That's one of the issue mentioned in https://github.com/zephyrproject-rtos/zephyr/issues/70945

ndrs-pst commented:
Thanks for the feedback! Please correct me if I'm wrong, the usage of `BT_SHELL` mostly resides with the `SERIAL` or `RTT` backend. Even though it's possible, I'm not sure about the use case for using `MQTT` or `TELNET` to remotely access `BT_SHELL`.

---
As for the previous question, I believe we can initialize it to `backend_name`, which is in `CONFIG_BT_SHELL_BACKEND`, by utilizing `shell_backend_get_by_name("backend_name")`. Even though we can have one or more shells, why can't we specify unsolicited output to just one specific backend?" 🤔

jhedberg commented on subsys/bluetooth/host/shell/bt_shell_private.c:
```diff
@@ -0,0 +1,53 @@
+
+#include "bt_shell_private.h"
+
+extern const struct shell *ctx_shell;
```
Wouldn't `bt_shell` be more appropriate, since this is now basically an internal API?

ndrs-pst replied:
That's also fine!
FWIW, I borrowed the name from `#include "net_shell_private.h"`, so I was following that pattern. 😄

Thalley commented:
> As for the previous question, I believe we can initialize it to `backend_name`, which is in `CONFIG_BT_SHELL_BACKEND`, by utilizing `shell_backend_get_by_name("backend_name")`. Even though we can have one or more shells, why can't we specify unsolicited output to just one specific backend?"

I'm not an expert in the shells, but the current `ctx_shell` is not a backend, but just a reference to a `struct shell`. Multiple `struct shell`s can use the same backend as far as I can tell, so not sure how you are expecting to use a specific backend. 

If you have 2 or more shells that each may call the BT shell command handler, then currently the `ctx_shell` only refers to one at them at any one time and may chance between them whenever `ctx_shell` is reassigned. This is clearly not ideal. 
Question is whether, and how, doing command X on shell S1 that trigger callback Y should send the string from the callback to just S1 or also S2 (another shell instance)?

ndrs-pst commented:
> If you have 2 or more shells that each may call the BT shell command handler, then currently the `ctx_shell` only refers to one at them at any one time and may chance between them whenever `ctx_shell` is reassigned. This is clearly not ideal.

Yeah, that's why I'm bringing up the real use case. Even if we have two or more shells, do we really need to support BT shell responses for every situation? It feels like we're trying to cover cases that might not happen often. 😃

> just S1 or also S2 (another shell instance)?

As you mentioned as an alternative in #70945. A kind of `shell_wall_print` may look like this

---

```
void bt_shell_wall_print_impl(const char *fmt, ...)
{
	va_list args;
	int count;
	const struct shell *sh;

	va_start(args, fmt);
	count = shell_backend_count_get();
	for (int i = 0; i < count; i++) {
	    va_list args_copy;

	    va_copy(args_copy, args);   /* Create a copy of 'args' to safely reuse */
	    sh = shell_backend_get(i);
	    shell_vfprintf(sh, SHELL_NORMAL, fmt, args_copy);

	    va_end(args_copy);          /* Clean up to prevent resource leaks */
	}
	va_end(args);
}
```

I'm OK if that's the way we can move forward.
Since the priority of this PR is to reduce the footprint and get closer to getting rid of `ctx_shell`.

**WDYT?**

Thalley commented:
> > If you have 2 or more shells that each may call the BT shell command handler, then currently the `ctx_shell` only refers to one at them at any one time and may chance between them whenever `ctx_shell` is reassigned. This is clearly not ideal.
> 
> Yeah, that's why I'm bringing up the real use case. Even if we have two or more shells, do we really need to support BT shell responses for every situation? It feels like we're trying to cover cases that might not happen often. 😃

It's hard to tell how people use these APIs, but the API does allow for such behavior, and thus any solution to our "problems" should consider what's possible, even if it will only happen with madmen :D 

> As you mentioned as an alternative in #70945. A kind of `shell_wall_print` may look like this
> 
> ```
> void bt_shell_wall_print_impl(const char *fmt, ...)
> {
> 	va_list args;
> 	int count;
> 	const struct shell *sh;
> 
> 	va_start(args, fmt);
> 	count = shell_backend_count_get();
> 	for (int i = 0; i < count; i++) {
> 	    va_list args_copy;
> 
> 	    va_copy(args_copy, args);   /* Create a copy of 'args' to safely reuse */
> 	    sh = shell_backend_get(i);
> 	    shell_vfprintf(sh, SHELL_NORMAL, fmt, args_copy);
> 
> 	    va_end(args_copy);          /* Clean up to prevent resource leaks */
> 	}
> 	va_end(args);
> }
> ```
> 
> I'm OK if that's the way we can move forward. Since the priority of this PR is to reduce the footprint and get closer to getting rid of `ctx_shell`.
> 
> **WDYT?**

I think the `wall` implementation looks fine :) @alwa-nordic was the one who originally suggested that idea, so maybe he has some input here too

ndrs-pst commented:
@alwa-nordic Thanks for the comment.
I'll address `host/shell` first and leave the others, e.g. `mesh/shell` and `audio/shell`, as they are.

ndrs-pst committed:
bluetooth: shell: replace `ctx_shell` with `sh` where applicable

Limit the usage of `ctx_shell` to cases where printing requires it
and `sh` is not available.

Signed-off-by: Pisit Sawangvonganan <pisit@ndrsolution.com>

ndrs-pst committed:
bluetooth: shell: refactor shell print to eliminate `ctx_shell` usage

This change aims to eliminate the dependency on `ctx_shell` in
the Bluetooth `host/shell/*`, making the code more maintainable.
Replaced `shell_*` functions that depended on `ctx_shell` with
the appropriate `bt_shell_*` functions.

The shell-less functions `bt_do_scan_filter_clear_name`, `bt_do_scan_off`,
and `bt_do_connect_le` were added so they can be called without `sh`.

Signed-off-by: Pisit Sawangvonganan <pisit@ndrsolution.com>

ndrs-pst commented:
Here we go! Finally, this PR is re-opened for review. 😃
Ping @Thalley, @jhedberg, and @alwa-nordic.

Thalley commented on subsys/bluetooth/host/shell/bt_shell_private.c:
```diff
@@ -0,0 +1,101 @@
+	}
+}
+
+void bt_shell_hexdump(const uint8_t *data, size_t len)
+{
+	int count;
+	const struct shell *sh;
+
+	count = shell_backend_count_get();
+	for (int i = 0; i < count; i++) {
+		sh = shell_backend_get(i);
+		shell_hexdump(sh, data, len);
+	}
+}
+
+void bt_shell_help(void)
+{
+	int count;
+	const struct shell *sh;
+
+	count = shell_backend_count_get();
+	for (int i = 0; i < count; i++) {
+		sh = shell_backend_get(i);
+		shell_help(sh);
+	}
+}
```
Since these 2 prints on all shells, should they have `wall` in the name too?

ndrs-pst replied:
Since the intention was to replace the `shell_*` functions that depended on `ctx_shell` with the appropriate `bt_shell_*` functions (in-short just adding `bt_` prefix) so I think it's no need to have `wall` in the name. :)

Thalley replied:
That seems to be inconsistent with `bt_shell_wall_vfprintf` which servers the same purpose, but has `wall` in the name, right?

ndrs-pst replied:
|shell (variant)|bt_shell (variant) (before)|bt_shell (variant) (opt 1)|bt_shell (variant) (opt 2)|
|---|---|---|---|
|shell_fprintf   |bt_shell_fprintf   |bt_shell_fprintf   |bt_shell_**wall**_fprintf   |
|shell_fprintf_info|bt_shell_fprintf_info   |bt_shell_fprintf_info   |bt_shell_**wall**_fprintf_info   |
|shell_fprintf_print|bt_shell_fprintf_print   |bt_shell_fprintf_print   |bt_shell_**wall**_fprintf_print   |
|shell_fprintf_warn |bt_shell_fprintf_warn|bt_shell_fprintf_warn|bt_shell_**wall**_fprintf_warn|
|shell_fprintf_error|bt_shell_fprintf_error|bt_shell_fprintf_error|bt_shell_**wall**_fprintf_error|
|shell_hexdump|bt_shell_hexdump|bt_shell_**wall**_hexdump|bt_shell_**wall**_hexdump|
|shell_help|bt_shell_help|bt_shell_**wall**_help|bt_shell_**wall**_help|
|shell_info|bt_shell_info|bt_shell_info|bt_shell_**wall**_info|
|shell_print|bt_shell_print|bt_shell_print|bt_shell_**wall**_print|
|shell_warn|bt_shell_warn|bt_shell_warn|bt_shell_**wall**_warn|
|shell_error|bt_shell_error|bt_shell_error|bt_shell_**wall**_error|

---

Just bring out the table of print functions provided by `bt_shell_private.h`.
Which option are we going to choose: opt1, opt2, or something else?

Thalley replied:
Since they are all `wall` in the end, then either the current, or opt 2. 

The main issue I had is that we had a single function using the `wall` name which seemed inconsistent as they are all `wall` variants. 

Of course since `bt_shell_wall_vfprintf` was an internal/static function not exposed by the header, it is less of any issue. 

So I'm OK with the current version or opt 2

ndrs-pst replied:
Thanks for the insight! I'll go with the current version. 🫡

Thalley reviewed (commented):
Looks good. Since BT is pretty much the only shell to do this, I think it makes sense to keep it in BT. One comment/question

Thalley commented:
Found one other place where "ctx_shell" was used outside of BT, but that was easy to just remove: https://github.com/zephyrproject-rtos/zephyr/pull/81395

Thalley commented on subsys/bluetooth/host/shell/bt_shell_private.h:
```diff
@@ -0,0 +1,99 @@
+ */
+void bt_shell_hexdump(const uint8_t *data, size_t len);
+
+/**
+ * @brief Prints the current command help.
+ * (Bluetooth context specific)
+ *
+ * Function will print a help string with: the currently entered command
+ * and subcommands (if they exist).
+ */
+void bt_shell_help(void);
```
`bt_shell_help` should only ever use the shell that tried to do something. I see that we have a case where `cmd_connect_le` is called from a callback. 

In that case (where `sh == NULL`), I would just not print this at all :) It's pretty useless in a callback (and with `wall`), since it related to a specific command.

ndrs-pst replied:
So, `bt_shell_help()` shouldn't be implemented, right? :)

Thalley replied:
I would say so, yeah. It's not useful to wall user errors :)

Thalley commented on subsys/bluetooth/host/shell/bt_shell_private.c:
```diff
@@ -0,0 +1,101 @@
+#include <zephyr/shell/shell_backend.h>
+#include "bt_shell_private.h"
+
+void bt_shell_wall_vfprintf(enum shell_vt100_color color, const char *fmt, va_list args)
```
```suggestion
static void wall_vfprintf(enum shell_vt100_color color, const char *fmt, va_list args)
```

ndrs-pst replied:
Got it will address on the next push

Thalley commented on subsys/bluetooth/host/shell/bt.c:
```diff
@@ -597,13 +597,13 @@ static void scan_recv(const struct bt_le_scan_recv_info *info, struct net_buf_si
 
 				/* "connect" is what would be in argv[0] normally */
-				cmd_connect_le(ctx_shell, 1, (char *[]){"connect"});
+				cmd_connect_le(NULL, 1, (char *[]){"connect"});
```
I think it's troublesome that we call these shell command handlers with a NULL shell. 

I fully understand that without the `ctx_shell`, we do not have a shell to use in those cases, but I think that we may need to refactor these calls in a way that they do not require a shell then, i.e. move the functionality from `cmd_connect_le` in a new function that is shell-less, and then call that function both here and in `cmd_connect_le`. 

Otherwise we end up in a situation where we cannot always assume that `sh != NULL`, and then we need to add `sh == NULL` checks everywhere. 

This will also remove the need for the `bt_shell_help` function

Thalley replied:
Yeah, this looks exactly what I hoped for :)

Thalley commented on subsys/bluetooth/host/shell/bt.c:
```diff
@@ -595,15 +596,17 @@ static void scan_recv(const struct bt_le_scan_recv_info *info, struct net_buf_si
 
 				/* "name" is what would be in argv[0] normally */
-				cmd_scan_filter_clear_name(ctx_shell, 1, (char *[]){"name"});
+				cmd_scan_filter_clear_name(NULL, 1, (char *[]){"name"});
```
Ditto; this should be a function so we don't call command handlers with a `NULL` sh

ndrs-pst replied:
Got it

ndrs-pst commented on subsys/bluetooth/host/shell/bt.c:
```diff
@@ -227,18 +229,18 @@ static void active_scan_timeout(struct k_work *work)
 		auto_connect.connect_name = false;
 		/* "name" is what would be in argv[0] normally */
-		cmd_scan_filter_clear_name(ctx_shell, 1, (char *[]){ "name" });
+		cmd_scan_filter_clear_name(NULL, 1, (char *[]){ "name" });
```
Got it

ndrs-pst commented:
Changes made as requested—hope this looks good!

Thalley commented:
Excellent job and great that you updated it so fast :)

ndrs-pst commented on subsys/bluetooth/host/shell/bt_shell_private.c:
It's true that this C-file isn't Bluetooth-specific, but since it's designed to address a Bluetooth-related problem, keeping it focused on Bluetooth is fine for now. If similar needs arise elsewhere, it can be refactored into a generic API later.

🤞 Once this is merged, I'll open another PR for `audio/shell`, `mesh/shell`, and `classic/shell` too.

jhedberg replied:
@jakub-uC @carlescufi any comments on this? How easily could this generic API be accepted as part of the shell subsystem?

jhedberg replied:
@jakub-uC @carlescufi I'm still waiting for some response to the above question.

Thalley replied:
Since the `ctx_shell` is only used in BT context, I think we can move forward with this PR. We can later port these to a non-BT place if there's any use for it

Thalley commented:
@ndrs-pst I think this PR actually solves https://github.com/zephyrproject-rtos/zephyr/issues/41796 - Should we add `fixes https://github.com/zephyrproject-rtos/zephyr/issues/41796` in the PR description here?

jhedberg commented on subsys/bluetooth/host/shell/bt.c:
```diff
@@ -1783,9 +1788,9 @@ static int cmd_scan_filter_set_addr(const struct shell *sh, size_t argc,
-			shell_error(ctx_shell,
-					"Invalid address string: %s\n",
-					addr_arg);
+			shell_error(sh,
+				    "Invalid address string: %s\n",
+				    addr_arg);
```
Can't this be just one line?

jhedberg commented on subsys/bluetooth/host/shell/bt.c:
```diff
@@ -125,8 +126,8 @@ static void print_le_addr(const char *desc, const bt_addr_le_t *addr)
 
-	shell_print(ctx_shell, "%s address: %s (%s)", desc, addr_str,
-		    addr_desc);
+	bt_shell_print("%s address: %s (%s)", desc, addr_str,
+		       addr_desc);
```
Merge to one line?

jhedberg commented on subsys/bluetooth/host/shell/bt.c:
```diff
@@ -396,14 +395,14 @@ static bool data_verbose_cb(struct bt_data *data, void *user_data)
-			shell_fprintf(ctx_shell, SHELL_WARNING,
-				      "BT_DATA_SVC_DATA32 data length too short (%u)",
-				      data->data_len);
+			bt_shell_fprintf_warn(
+				"BT_DATA_SVC_DATA32 data length too short (%u)",
+				data->data_len);
```
Seems like at least the string literal should be mergeable to the first line

jhedberg commented on subsys/bluetooth/host/shell/bt.c:
```diff
@@ -418,22 +417,22 @@ static bool data_verbose_cb(struct bt_data *data, void *user_data)
-			shell_fprintf(ctx_shell, SHELL_WARNING,
-				      "BT_DATA_SVC_DATA128 data length too short (%u)",
-				      data->data_len);
+			bt_shell_fprintf_warn(
+				"BT_DATA_SVC_DATA128 data length too short (%u)",
+				data->data_len);
```
Same here

jhedberg reviewed (commented):
Some style comments. I'm still a bit concerned with the total silence from the shell maintainers.

Thalley commented on subsys/bluetooth/host/shell/bt.c:
```diff
@@ -857,8 +856,8 @@ static void identity_resolved(struct bt_conn *conn, const bt_addr_le_t *rpa,
 
-	shell_print(ctx_shell, "Identity resolved %s -> %s", addr_rpa,
-	      addr_identity);
+	bt_shell_print("Identity resolved %s -> %s", addr_rpa,
+		       addr_identity);
```
Hmm, the way that it's formatted looks like clang-format, but where the line length is set to 80 instead of 100. Chances are that the author is using a misconfigured auto-formatter

ndrs-pst replied:
As far as I can recall, I used the in-tree `.clang-format` with some manual adjustments to align it with the previous one.
For the next push, I'll change it according to the clang-format suggestion.

ndrs-pst commented:
> @ndrs-pst I think this PR actually solves #41796 - Should we add `fixes https://github.com/zephyrproject-rtos/zephyr/issues/41796` in the PR description here?

Got it.

jhedberg reviewed (approved):
Approving, but there's potential cleanup still to be done in follow-up PRs (see my inline comments). Please also engage actively with the shell maintainers to make this a generic API eventually. Thanks.

Thalley commented on subsys/bluetooth/host/shell/bt.c:
```diff
@@ -833,18 +824,18 @@ static void disconnected(struct bt_conn *conn, uint8_t reason)
-		    param->latency, param->timeout);
+	bt_shell_print("LE conn  param req: int (0x%04x, 0x%04x) lat %d"
+		       " to %d", param->interval_min, param->interval_max,
+		       param->latency, param->timeout);
 
 	return true;
 }
 
 static void le_param_updated(struct bt_conn *conn, uint16_t interval,
 			     uint16_t latency, uint16_t timeout)
 {
-	shell_print(ctx_shell, "LE conn param updated: int 0x%04x lat %d "
-		     "to %d", interval, latency, timeout);
+	bt_shell_print("LE conn param updated: int 0x%04x lat %d "
+		       "to %d", interval, latency, timeout);
```
Agreed. If the author has setup clang-format in their IDE with only formatting changed lines using the .clang-format from Zephyr, these would have been formatted like you want.

ndrs-pst commented:
Thank you for the approval and feedback!
I'll address the potential cleanup you mentioned in a follow-up PR.
Cheers 🎉
